components/org.wso2.micro.integrator.initializer/src/main/java/org/wso2/micro/integrator/initializer/handler/transaction/TransactionCountHandler.java

```diff
@@ -69,28 +71,56 @@ public static long getTransactionCount() {
     }
 
     /**
-     * If the current transaction count has reached the Max Integer value, commit the current transaction to the
-     * database and reset the transaction count to zero.
+     * Increment the transaction count by 1 if the INTERNAL_TRANSACTION_COUNTED message property is not set to true.
+     * Commit the current transaction count to database and set the transaction count to zero if it has exceeded the
+     * Integer.MAX_VALUE
+     */
+    private void handleTransactionCount(MessageContext messageContext) {
+        org.apache.axis2.context.MessageContext axis2MessageContext =
+                ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+        Object transactionProperty = axis2MessageContext.getProperty(BaseConstants.INTERNAL_TRANSACTION_COUNTED);
+        // increment the transaction count by 1 if the INTERNAL_TRANSACTION_COUNTED message property has not been set
+        // or if the value of the property is set to false.
+        if (!(transactionProperty instanceof Boolean && (Boolean) transactionProperty)) {
+            axis2MessageContext.setProperty(BaseConstants.INTERNAL_TRANSACTION_COUNTED, true);
+            // commit the current transaction count to database and set the transaction count to zero if it has
+            // reached to Integer.MAX_VALUE
+            synchronized (lock) {
+                commitAndResetTransactionCountIfReachedToMaxIntegerValue();
+                transactionCount += 1;
+            }
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Transaction " + axis2MessageContext.getLogCorrelationID() + " is counted.");
+            }
+        } else {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Skipped counting the transaction " + axis2MessageContext.getLogCorrelationID()
+                                  + " since it has been counted before.");
+            }
+        }
+    }
+
+    /**
+     * If the current transaction count has reached or exceeded the Max_Integer_value * 0.9, commit the current
+     * transaction to the database, set a new node id, and reset the transaction count to zero.
      */
     private void commitAndResetTransactionCountIfReachedToMaxIntegerValue() {
-        if (transactionCount == Integer.MAX_VALUE) {
+        if (transactionCount >= Integer.MAX_VALUE * 0.9) {
```

Comment:
Shall we assign this (Integer.MAX_VALUE * 0.9) to a const and use it here

Reply:
Done.